core/src/main/java/org/apache/spark/memory/TaskMemoryManager.java

```diff
@@ -101,27 +106,97 @@
    */
   public TaskMemoryManager(MemoryManager memoryManager, long taskAttemptId) {
     this.inHeap = memoryManager.tungstenMemoryIsAllocatedInHeap();
     this.memoryManager = memoryManager;
     this.taskAttemptId = taskAttemptId;
+    this.consumers = new HashMap<>();
   }
 
   /**
-   * Acquire N bytes of memory for execution, evicting cached blocks if necessary.
+   * Acquire N bytes of memory for a consumer. If there is no enough memory, it will call
+   * spill() of consumers to release more memory.
+   *
    * @return number of bytes successfully granted (<= N).
    */
-  public long acquireExecutionMemory(long size) {
-    return memoryManager.acquireExecutionMemory(size, taskAttemptId);
+  public long acquireExecutionMemory(long size, MemoryConsumer consumer) throws IOException {
+    assert(size >= 0);
+    synchronized (this) {
+      long got = memoryManager.acquireExecutionMemory(size, taskAttemptId);
+
+      // call spill() on itself to release some memory
+      if (got < size && consumer != null) {
+        consumer.spill(size - got, consumer);
+        got += memoryManager.acquireExecutionMemory(size - got, taskAttemptId);
+      }
+
+      if (got < size) {
+        long needed = size - got;
+        // call spill() on other consumers to release memory
+        for (MemoryConsumer c: consumers.keySet()) {
+          if (c != null && c != consumer) {
+            needed -= c.spill(size - got, consumer);
+            if (needed < 0) {
+              break;
+            }
+          }
+        }
+        got += memoryManager.acquireExecutionMemory(size - got, taskAttemptId);
+      }
+
+      long old = 0L;
+      if (consumers.containsKey(consumer)) {
+        old = consumers.get(consumer);
+      }
+      consumers.put(consumer, got + old);
+
+      return got;
+    }
   }
 
   /**
-   * Release N bytes of execution memory.
+   * Release N bytes of execution memory for a MemoryConsumer.
    */
-  public void releaseExecutionMemory(long size) {
-    memoryManager.releaseExecutionMemory(size, taskAttemptId);
+  public void releaseExecutionMemory(long size, MemoryConsumer consumer) {
+    assert(size >= 0);
+    if (size == 0) {
+      return;
+    }
+    synchronized (this) {
+      if (consumers.containsKey(consumer)) {
+        long old = consumers.get(consumer);
+        if (old > size) {
+          consumers.put(consumer, old - size);
+        } else {
+          if (old < size) {
+            if (Utils.isTesting()) {
+              Platform.throwException(
+                new SparkException("Release more memory " + size + "than acquired " + old + " for "
+                  + consumer));
+            } else {
+              logger.warn("Release more memory " + size + " than acquired " + old + " for "
+                + consumer);
+            }
+          }
+          consumers.remove(consumer);
+        }
+      } else {
+        if (Utils.isTesting()) {
+          Platform.throwException(
```

Comment:
Maybe add new method to Utils, accepting message String, which covers lines 191 to 196